APIv4 - Add SortableEntity type which auto-adjusts weights

colemanw opened the pull request:
Overview
----------------------------------------
Gives APIv4 the ability to manage the "weight" of certain entities.

Before
----------------------------------------
- Saving a new record in APIv4 without specifying a weight had undefined behavior
- Trying to re-order a list was difficult - all items had to be manually re-adjusted

After
----------------------------------------
- APIv4 will supply an auto-increment default value for weights
- Simply by updating the weight of one record, others will auto-adjust to make room for it.

Technical Details
----------------------------------------
This creates a trait for a new type: "SortableEntity", as well as a new `@orderBy` annotation which specifies the name of the column containing sorting data.

demeritcowboy commented on CRM/Utils/Weight.php:
```diff
@@ -154,12 +151,18 @@ public static function updateOtherWeights($daoName, $oldWeight, $newWeight, $fie
+    $existing = self::query('SELECT', $daoName, $fieldValues, "id", "$weightField = $newWeight");
+    // Nothing to do if no existing record has this weight
+    if (empty($existing->N)) {
+      return $newWeight;
```
I'm not 100% sure about this because there's a couple places where the comments indicate it's supposed to create a gap for the new row, and there are several calls to this function. But logically the gap would already be there if it doesn't exist. I tried clicking around a couple places and it seems ok.

I'm not sure how to evaluate the api/search kit changes since I'm not sure where they're used. So have given merge-ready.

colemanw replied:
@demeritcowboy I think it was an oversight. Prior to this PR there were no unit tests for `CRM_Utils_Weight`.

colemanw commented:
@demeritcowboy thanks for the review. I've just rebased this against master and added a few more unit tests + fixes to make sure `SortableEntity` and `ManagedEntity` traits are working well together in an export.

demeritcowboy commented:
Ha ha sneaky - a lot more changes now. Ok will take a look.

colemanw commented:
Yea, I keep adding unit tests and it surfaces subtle bugs. It all came together in the `Navigation` entity which is sortable, managed, and domain-specific. So I needed to make the `Export` action filter by domain and order by weight.

demeritcowboy commented:
I clicked around a bit and I think this is ok, just there's a lot going on in the api4 side that I don't have a good sense of what it affects, so wouldn't mind someone just doing a quick once-over.

I notice searchkit for option values where you add a where for option_group_id returns nothing, but it's not from this PR.

eileenmcnaughton commented on CRM/Core/BAO/ConfigSetting.php:
```diff
@@ -321,14 +321,23 @@ public static function enableComponent($componentName) {
+   */
+  public static function enableAllComponents() {
+    $allComponents = array_keys(CRM_Core_Component::getComponents());
+    if (Civi::settings()->get('enable_components') != $allComponents) {
```
Can we use !==  here

colemanw replied:
![image](https://user-images.githubusercontent.com/2874912/143947592-f8bc3abc-7a1f-4ce3-a1ac-bd8a18eac424.png)

Here we are checking that both arrays contain the same elements, but they don't have to be in the same order.

demeritcowboy replied:
Hmm, interesting:
```
php -r '$a = array(1,2); $b = array(2,1); var_dump($a == $b); var_dump($a === $b);'
bool(false)
bool(false)
```
(php 7.4 at least)

colemanw replied:
I recall stepping through this in the debugger and confirming that the condition is true the first pass but false subsequent test iterations, but wasn't looking too closely.

eileenmcnaughton commented on CRM/Core/OptionGroup.php:
```diff
@@ -21,13 +21,23 @@ class CRM_Core_OptionGroup {
    * $_domainIDGroups array maintains the list of option groups for whom
    * domainID is to be considered.
+   *
+   * FIXME: Hardcoded list = bad. It would be better to make this a column in the civicrm_option_group table
```
I've always wondered about declaring our option groups in files (like with settings). It would take some work to do but I think it would be easier to maintain than the current way we manage the installation & could be hookable etc

eileenmcnaughton commented:
Ouch - yeah - it made my head hurt a bit too - but I didn't spot anything of concern

colemanw commented:
Thanks @eileenmcnaughton and @demeritcowboy!
